### PR TITLE
feat(dag): implement GEDF

### PR DESCRIPTION
**Pull request by nokosaaan**
## Description
・Implemented the GEDF scheduler for DAGs
・Modified existing implementations: awkernel_async_lib/src/dag.rs and awkernel_async_lib/src/task.rs
・Partially revised specifications around spawn operations
・No visual changes to user-written code

## Related links
[Google Slide](https://docs.google.com/presentation/d/e/2PACX-1vSKQFr6JN5wRSocGAhF1E2dfvP-01M245CKbm-a5IaJpZsQibXKXbdgvZZRoxfD4A/pub?start=false&loop=false&delayms=3000) (gave a presentation at the September 5th TIER IV-CREST meeting)

## How was this PR tested?
・[raspi4](https://github.com/tier4/awkernel?tab=readme-ov-file#raspberry-pi-4-aarch64)
・test logs below
Welcome to Awkernel!

You can use BLisp language as follows.
https://ytakano.github.io/blisp/

> (factorial 20)
2432902008176640000
> (+ 10 20)
30

Enjoy!

> [        11206 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:18: period is 1000000000 [ns]
[        11206 INFO] DAG created successfully
[        11306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        11315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        11315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        11325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        11336 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        12306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        12306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        12306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        12315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        12325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        13306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        13306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        13306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        13315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        13325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        14306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        14306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        14306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        14306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        14306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        15306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        15306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        15306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        15306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        15306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        16306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        16306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        16306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        16306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        16306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        17306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        17306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        17306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        17306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        17306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        18306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        18306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        18306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        18306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        18306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        19306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        19306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        19306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        19306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        19306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        20306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        20306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        20306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        20315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        20325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        21306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        21306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        21306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        21306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        21306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        22306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        22306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        22306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        22315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        22325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        23306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        23306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        23306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        23306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        23306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        24306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        24306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        24306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        24315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        24325 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        25306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        25306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        25306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        25306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        25306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        26306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        26306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        26306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        26315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        26315 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        27306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        27306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        27306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        27306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        27306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        28306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        28306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        28306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        28306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        28306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        29306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        29306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        29306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        29306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        29306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4
[        30306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:28: value=1 in reactor_source_node
[        30306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:43: value=2 in reactor_node1
[        30306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:59: value=4 in reactor_node2
[        30306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:74: value=9 in reactor_node3
[        30306 DEBUG] /home/nokosan/azumi-lab/awkernel3/applications/tests/test_dag/src/lib.rs:89: value=13 in reactor_node4

## Notes for reviewers
Instead of `test_gedf` , I conducted the test by changing the scheduler type of `test_dag` to `GEDF(0)`. The test logs are as shown above. **test logs updated on September 25th.**

### CODE REVIEW

File: awkernel_async_lib/src/dag.rs [hunk 1]
```diff
@@ -463,6 +510,9 @@ struct NodeInfo {
     relative_deadline: Option<Duration>,
 }
 
+pub fn to_node_index(index: u32) -> NodeIndex {
+    NodeIndex::new(index as usize)
+}
```
Comment (kobayu858) [inline awkernel_async_lib/src/dag.rs:515]: Please insert a blank line between EdgeInfo.

File: awkernel_async_lib/src/task.rs [hunk 2]
```diff
@@ -352,7 +365,58 @@ pub fn spawn(
...
+/// let deadline = Duration::from_millis(100);
+/// set_relative_deadline_public(&dag, sink_node_idx, deadline);
+/// let task_id = task::spawn_with_dag_info(
+///     "dag task".into(),
+///     async { Ok(()) },
+///     SchedulerType::GEDF(0),
+///     Some((1, 0))  // dag_info as Option<(u32, u32)>
+/// );
+/// ```
+pub fn spawn_with_dag_info(
```
Comment (kobayu858) [inline awkernel_async_lib/src/task.rs:399]: Since there is significant overlap with the spawn function, please use helper functions to unify the code.

For example, imagine `spawn` and `spawn_with_dag_info` as interfaces only, with the actual implementation handled by `spawn_inner`.

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 3]
```diff
@@ -71,10 +72,48 @@ impl Scheduler for GEDFScheduler {
...
+
+                    if let Some((dag_id, node_index)) = dag_info {
+                        if get_dag_absolute_deadline(dag_id).is_none() {
+                            let dag =
+                                get_dag(dag_id).expect("GEDF scheduler: DAG {dag_id} not found");
+                            let relative_deadline_ms = dag
+                                .get_sink_relative_deadline()
+                                .map(|deadline| deadline.as_millis() as u64)
+                                .unwrap_or_else(|| {
+                                    panic!("GEDF scheduler: DAG {dag_id} has no sink relative deadline set");
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/scheduler/gedf.rs:89]: This syntax is correct.

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 4]
```diff
@@ -71,10 +72,48 @@ impl Scheduler for GEDFScheduler {
...
+                                get_dag(dag_id).expect("GEDF scheduler: DAG {dag_id} not found");
+                            // Use `to_node_index` to convert the u32 type to NodeIndex
+                            // for use with the method that determines whether it is a source node.
+                            let current_node_index = to_node_index(node_index);
+                            let is_source_node = dag.is_source_node(current_node_index);
+                            if is_source_node {
+                                let relative_deadline_ms = dag.get_sink_relative_deadline()
+                                    .map(|deadline| deadline.as_millis() as u64)
+                                    .unwrap_or_else(|| {
+                                        panic!("GEDF scheduler: DAG {dag_id} has no sink relative deadline set");
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/scheduler/gedf.rs:104]: This syntax is correct.

File: awkernel_async_lib/src/dag.rs [hunk 5]
```diff
@@ -525,6 +576,35 @@ pub fn get_dag(id: u32) -> Option<Arc<Dag>> {
...
+pub fn set_dag_absolute_deadline(dag_id: u32, deadline: u64) -> bool {
+    if let Some(dag) = get_dag(dag_id) {
+        dag.set_absolute_deadline(deadline);
+        true
+    } else {
+        false
+    }
+}
+
+// for test
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/dag.rs:594]: This section will be deleted.

File: awkernel_async_lib/src/dag.rs [hunk 6]
```diff
@@ -525,6 +576,35 @@ pub fn get_dag(id: u32) -> Option<Arc<Dag>> {
...
+pub fn add_node_with_topic_edges_public(
+    dag: &Arc<Dag>,
+    subscribe_topic_names: &[Cow<'static, str>],
+    publish_topic_names: &[Cow<'static, str>],
+) -> NodeIndex {
+    dag.add_node_with_topic_edges(subscribe_topic_names, publish_topic_names)
+}
+
+//for test
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/dag.rs:603]: This section will be deleted.

File: awkernel_async_lib/src/dag.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/dag.rs:594]: This section will be deleted.

File: awkernel_async_lib/src/dag.rs [hunk 7]
```diff
@@ -525,6 +576,35 @@ pub fn get_dag(id: u32) -> Option<Arc<Dag>> {
...
+pub fn set_relative_deadline_public(dag: &Arc<Dag>, node_idx: NodeIndex, deadline: Duration) {
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/dag.rs:604]: This section will be deleted.

File: awkernel_async_lib/src/task.rs [hunk 8]
```diff
@@ -352,7 +365,58 @@ pub fn spawn(
...
+/// let sink_node_idx = add_node_with_topic_edges_public(&dag, &[], &[]);
+/// let deadline = Duration::from_millis(100);
+/// set_relative_deadline_public(&dag, sink_node_idx, deadline);
+/// let task_id = task::spawn_with_dag_info(
+///     "dag task".into(),
+///     async { Ok(()) },
+///     SchedulerType::GEDF(0),
+///     Some((1, 0))  // dag_info as Option<(u32, u32)>
+/// );
+/// ```
```
[the remark this answers is not in the document]
Comment (kobayu858) [reply awkernel_async_lib/src/task.rs:398]: [Ref: ](https://github.com/tier4/awkernel/blob/main/awkernel_async_lib/src/time_interval.rs#L88-L101)

**Comment by kobayu858** [comment]
@nokosaaan 
Please add GEDF test cases, as none currently exist. The current `test_gedf` is intended for regular tasks.

Also, please copy and paste the test logs into the “How was this PR tested?” section.

File: awkernel_async_lib/src/dag.rs (hunk 1 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/dag.rs:516]: inserted a blank line

File: awkernel_async_lib/src/task.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nokosaaan) [reply awkernel_async_lib/src/task.rs:377]: I added `ignore` to prevent running docs tests in `spawn_with_dag_info`. I also believe this change is acceptable since the docs tests are being run during `spawn`.

File: awkernel_async_lib/src/task.rs (hunk 2 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/task.rs:378]: To consolidate duplicate processing in `spawn` and `spawn_with_dag_info`, I added `spawn_inner`.

File: awkernel_async_lib/src/task.rs [hunk 9]
```diff
@@ -272,6 +283,8 @@ impl Tasks {
                     need_sched: false,
                     need_preemption: false,
                     panicked: false,
+                    dag_id: dag_info.map(|(d, _)| d),
+                    node_index: dag_info.map(|(_, n)| n),
```
[the remark this answers is not in the document]
Comment (nokosaaan) [reply awkernel_async_lib/src/task.rs]: I designed the `DagInfo` structure to be flexible enough to accommodate future changes. By wrapping it in an Option type when necessary, I also enabled it to handle operations requiring option-type return values.

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 10]
```diff
@@ -71,10 +72,50 @@ impl Scheduler for GEDFScheduler {
...
             let mut node_inner = MCSNode::new();
             let mut info = task.info.lock(&mut node_inner);
+            let dag_info = info.get_dag_info();
             match info.scheduler_type {
                 SchedulerType::GEDF(relative_deadline) => {
                     let wake_time = awkernel_lib::delay::uptime();
-                    let absolute_deadline = wake_time + relative_deadline;
+                    let absolute_deadline;
+
+                    if let Some(dag_info) = dag_info {
```
Comment (kobayu858) [inline awkernel_async_lib/src/scheduler/gedf.rs:81]: The nested structure is difficult to read, so please use helper functions.

[Ref Site](https://qiita.com/DeployCat/items/1ec901864d4ab11c8d6f#%E3%81%A9%E3%81%AE%E7%A8%8B%E5%BA%A6%E3%81%BE%E3%81%A7%E3%81%AA%E3%82%89%E8%A8%B1%E5%AE%B9%E3%81%A7%E3%81%8D%E3%82%8B%E3%81%8B)

Comment (nokosaaan) [reply awkernel_async_lib/src/scheduler/gedf.rs]: I refactored the nested structure using helper functions and early return.

File: applications/tests/test_dag/src/lib.rs [hunk 11]
```diff
@@ -30,7 +30,7 @@ pub async fn run() {
             (number,)
         },
         vec![Cow::from("topic0")],
-        SchedulerType::PrioritizedFIFO(30),
+        SchedulerType::GEDF(0),
```
Comment (atsushi421) [inline applications/tests/test_dag/src/lib.rs:33]: The realtive deadline must be greater than zero.

File: awkernel_async_lib/src/task.rs [hunk 12]
```diff
@@ -240,6 +246,12 @@ struct Tasks {
     id_to_task: BTreeMap<u32, Arc<Task>>,
 }
 
+#[derive(Clone)]
+pub struct DagInfo {
+    pub dag_id: u32,
+    pub node_index: u32,
```
Comment (atsushi421) [inline awkernel_async_lib/src/task.rs:252]: Please unify to either `_id` or `_index`.

File: awkernel_async_lib/src/dag.rs [hunk 13]
```diff
@@ -190,6 +192,23 @@ impl Dag {
...
 
+    pub fn is_source_node(&self, node_index: NodeIndex) -> bool {
+        let source_nodes = self.get_source_nodes();
+        source_nodes.contains(&node_index)
+    }
+
+    pub fn get_sink_relative_deadline(&self) -> Option<Duration> {
+        let mut node = MCSNode::new();
+        let graph = self.graph.lock(&mut node);
+        let sink_nodes: Vec<NodeIndex> = graph.externals(Direction::Outgoing).collect();
```
Comment (atsushi421) [inline awkernel_async_lib/src/dag.rs:203]: Please use `get_sink_nodes()`.

File: awkernel_async_lib/src/dag.rs [hunk 14]
```diff
@@ -190,6 +192,23 @@ impl Dag {
...
+
+        if let Some(sink_node_index) = sink_nodes.first() {
```
Comment (atsushi421) [inline awkernel_async_lib/src/dag.rs:205]: Could you add a comment or assertion stating that this function returns the relative deadline of the first sink node?

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 15]
```diff
@@ -184,3 +191,37 @@ impl GEDFScheduler {
...
     }
 }
+
+pub fn calculate_and_update_dag_deadline(dag_info: &DagInfo, wake_time: u64) -> u64 {
+    let dag_id = dag_info.dag_id;
+    let node_index = dag_info.node_index;
+
+    if let Some(absolute_deadline) = get_dag_absolute_deadline(dag_id) {
+        let dag = get_dag(dag_id).expect("GEDF scheduler: DAG {dag_id} not found");
+        let current_node_index = to_node_index(node_index);
```
Comment (atsushi421) [inline awkernel_async_lib/src/scheduler/gedf.rs:201]: How about directly using `NodeIndex` type instead of `u32` for `DagInfo.node_index` without `to_node_index()`?

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 16]
```diff
@@ -71,10 +72,16 @@ impl Scheduler for GEDFScheduler {
...
             let mut node_inner = MCSNode::new();
             let mut info = task.info.lock(&mut node_inner);
+            let dag_info = info.get_dag_info();
             match info.scheduler_type {
                 SchedulerType::GEDF(relative_deadline) => {
                     let wake_time = awkernel_lib::delay::uptime();
-                    let absolute_deadline = wake_time + relative_deadline;
+                    let absolute_deadline = if let Some(ref dag_info) = dag_info {
+                        calculate_and_update_dag_deadline(dag_info, wake_time)
+                    } else {
```
Comment (atsushi421) [inline awkernel_async_lib/src/scheduler/gedf.rs:81]: Could you explain when this else branch is chosen?

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: awkernel_async_lib/src/scheduler/gedf.rs [hunk 17]
```diff
@@ -184,3 +191,37 @@ impl GEDFScheduler {
...
+        return dag_absolute_deadline;
+    }
+
+    let dag = get_dag(dag_id).expect("GEDF scheduler: DAG {dag_id} not found");
+    let relative_deadline_ms = dag
+        .get_sink_relative_deadline()
+        .map(|deadline| deadline.as_millis() as u64)
+        .unwrap_or_else(|| {
+            panic!("GEDF scheduler: DAG {dag_id} has no sink relative deadline set");
+        });
```
[the remark this answers is not in the document]
Comment (atsushi421) [reply awkernel_async_lib/src/scheduler/gedf.rs:223]: I think so too.

[the remark this answers is not in the document]
Comment (nokosaaan) [reply awkernel_async_lib/src/scheduler/gedf.rs]: I fixed in commit `fix: adding a function and code optimization`.

File: applications/tests/test_dag/src/lib.rs (hunk 11 above)
Comment (nokosaaan) [reply applications/tests/test_dag/src/lib.rs]: I fixed in commit `fix: adding a function and code optimization`.

File: awkernel_async_lib/src/dag.rs (hunk 14 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/dag.rs:204]: I added a comment.

File: awkernel_async_lib/src/scheduler/gedf.rs (hunk 16 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/scheduler/gedf.rs:81]: This GEDF scheduler also handles scheduling regular tasks. If a task lacks a `dag_info`, it is treated as a regular task. When the scheduler uses its `relative_deadline` to return an absolute deadline, it branches to this `else` clause.

File: awkernel_async_lib/src/task.rs (hunk 12 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/task.rs]: I fixed in commit [fix: unify the name index to id](https://github.com/tier4/awkernel/pull/653/commits/3e20931da064c0905bcaa56bc0b9dc2027ec259e).

File: awkernel_async_lib/src/scheduler/gedf.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (atsushi421) [reply awkernel_async_lib/src/scheduler/gedf.rs]: I think it's better to include these lines to helper function.

```rust
let dag_absolute_deadline = wake_time + relative_deadline_ms;
        set_dag_absolute_deadline(dag_id, dag_absolute_deadline);
```

File: awkernel_async_lib/src/scheduler/gedf.rs (hunk 16 above)
Comment (nokosaaan) [reply awkernel_async_lib/src/scheduler/gedf.rs:81]: I added a comment.

**Review by atsushi421 — APPROVED** [review]
LGtm
